include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,457 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_ASYNC_POOL_H_
+#define DALI_CORE_MM_ASYNC_POOL_H_
+
+#include <algorithm>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+#include "dali/core/mm/pool_resource.h"
+#include "dali/core/mm/detail/free_list.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/cuda_event_pool.h"
+#include "dali/core/cuda_stream.h"
+#include "dali/core/device_guard.h"
+
+namespace dali {
+namespace mm {
+
+template <memory_kind kind, class FreeList, class LockType, class Upstream = memory_resource<kind>>
+class async_pool_base : public stream_aware_memory_resource<kind> {
+ public:
+  /**
+   * @param upstream       Upstream resource, used by the global pool
+   * @param avoid_upstream If true, synchronize with outstanding deallocations before
+   *                       using upstream.
+   */
+  explicit async_pool_base(Upstream *upstream, bool avoid_upstream = true)
+  : global_pool_(upstream), avoid_upstream_(avoid_upstream) {
+  }
+  ~async_pool_base() {
+    try {
+      synchronize();
+    } catch (const CUDAError &e) {
+      if ((e.is_rt_api() && e.rt_error() != cudaErrorCudartUnloading) ||
+          (e.is_drv_api() && e.drv_error() != CUDA_ERROR_DEINITIALIZED))
+        std::terminate();
+    }
+  }
+
+  /**
+   * @brief Waits until all pending frees are finished.
+   */
+  void synchronize() {
+    DeviceGuard dg(device_id_);
+    synchronize_impl(true);
+  }
+
+ private:
+  void synchronize_impl(bool lock) {
+    {
+      std::unique_lock<std::mutex> ulock(lock_, std::defer_lock);
+      if (lock)
+        ulock.lock();
```

Comment:
Maybe you can move this lock to `synchronize()`?

Reply:
There would be too many layers to pass the lock through when calling synchronize from `allocate_from_global_pool`